### ISSUE TITLE
Setting a value of a descriptor in a class method does not report type errors and expands the attribute's type

### ISSUE THREAD

**Issue by svenberkvens** [microsoft/pyright#10249]
**Describe the bug**
When defining an attribute on a class using a descriptor (with which I mean a class with `__get__` and `__set__` methods), pyright displays some unexpected behavior: when setting a value for the attribute in a method of the class, no type error is raised if the wrong type is assigned, where the same assignment outside of the class definition **does** give a type error. What's more, besides accepting the incorrect type in the assignment, pyright also expands the type of the attribute to include the incorrect type.

**Code or Screenshots**

```python
from typing import Any, overload, reveal_type


class Descriptor:
    @overload
    def __get__(self, instance: None, owner: Any) -> type[int]:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> int:
        ...

    def __get__(self, instance, owner) -> type[int] | int:
        del owner
        if instance is None:
            return int
        return 1

    def __set__(self, instance: object, value: int) -> None:
        print(instance, value)

class Test:
    attr = Descriptor()

    def func(self):
        self.attr = "test"  # Why no error here? And why does this alter self.attr's type to include str?
        pass

# Reveals: type[int] | str, should have been type[int]
reveal_type(Test.attr)

t = Test()
# Reveals: int | str, should have been int
reveal_type(t.attr)

t.func()
t.attr = "test"  # Doing the same as in func() here DOES yield an error
```

**VS Code extension or command-line**
The above was checked with the command line version of pyright, version 1.1.398.

I also tried with a ancient version of pyright, for example 1.1.130, and it gives errors on the same source file locations, although the wording is different.

**Mypy comparison**
I also checked this on the Mypy playground, latest version, and it does infer the correct typing (on the `reveal_type` lines). It, however, also fails to report an error for the assignment in  the `func` function.

**Comment by svenberkvens** [comment]
This simpler version without overloads and stuff exhibits the same problems:

```python
from typing import reveal_type


class Descriptor:
    def __get__(self, instance, owner) -> int:
        del instance, owner
        return 1

    def __set__(self, instance: object, value: int) -> None:
        print(instance, value)

class Test:
    attr = Descriptor()

    def func(self):
        self.attr = "test"

reveal_type(Test.attr)
reveal_type(Test().attr)

Test().attr = "test"
```

**Comment by erictraut** [comment]
The issue here is that you haven't declared the type of `Test.attr`. Its type is therefore inferred based on assigned values within the implementation. If you declare its type (`attr: Descriptor = Descriptor()`), then pyright will enforce it.

Alternatively, if you declare `attr` as a class variable (`attr: ClassVar = Descriptor()`), pyright will then enforce its use as a class variable.

If you don't specify that `attr` is a class variable and supply no type declaration for it, then you can assign values to it at the class and instance level.

```python
class Test:
    @classmethod
    def cm(cls):
        cls.attr = Descriptor()
    
    def im(self):
        self.attr = ""
```

For an explanation of how pyright treats attributes, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=class-and-instance-variables). Mypy's behaviors here [differ somewhat](https://microsoft.github.io/pyright/#/mypy-comparison?id=class-and-instance-variable-inference).

In any case, pyright is working as intended here, so I don't consider this a bug.

**Comment by svenberkvens** [comment]
I beg to differ.

Setting a variable to a class instance **at definition time** that looks like a descriptor will make that name act like a descriptor. That name will then behave the way the descriptor indicates, unless you change the class. So I think only assignments to the class's notion of the name (not the class instance) should influence the type of the attribute, if that name is present in the class's definition and is defined as a descriptor there.

Explicit typing is certainly possible (and will help me well enough; thank you!), but I think pyright should follow the mechanics of the way in which Python handles descriptors when classes are defined: if a name is assigned to a descriptor-type object at definition time, then that's the way the name will behave.
